<a href="https://colab.research.google.com/github/ramzymhafeel/SE4050-Deep-Learning-Lab-01/blob/main/notebooks/SE4050_Lab01_Task2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SE4050 — Deep Learning

## Lab 01: Introduction to Python Libraries for Deep Learning

### Task 2 — Au Nanoparticle Dataset Analysis

**Student Name:** Mohamed Ramzy  
**Index Number:** IT23386822
**Module:** SE4050 — Deep Learning  
**Academic Year:** 2026

## Objective

The objective of Task 2 is to use Pandas, Matplotlib, and Seaborn to analyze and visualize selected features from the Au nanoparticle dataset.

The selected features are:

- `N_total`
- `N_bulk`
- `N_surface`
- `R_avg`

In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print("Required libraries imported successfully.")

Required libraries imported successfully.


In [2]:
from google.colab import drive

drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
dataset_folder = (
    "/content/drive/MyDrive/"
    "SE4050_DeepLearning/"
    "Lab_01/"
    "02_Datasets"
)

au_file_path = os.path.join(
    dataset_folder,
    "Au_nanoparticle_dataset.csv"
)

print("Dataset path:")
print(au_file_path)

Dataset path:
/content/drive/MyDrive/SE4050_DeepLearning/Lab_01/02_Datasets/Au_nanoparticle_dataset.csv


## Task 2.1 — Load the Au Nanoparticle Dataset

### Introduction

The Au nanoparticle dataset is loaded using Pandas for data analysis and visualization.

The analysis focuses on the following four numerical features:

- `N_total`
- `N_bulk`
- `N_surface`
- `R_avg`

In [4]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print("Required libraries imported successfully.")

Required libraries imported successfully.


In [5]:
from google.colab import drive

drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
dataset_folder = (
    "/content/drive/MyDrive/"
    "SE4050_DeepLearning/"
    "Lab_01/"
    "02_Datasets"
)

au_file_path = os.path.join(
    dataset_folder,
    "Au_nanoparticle_dataset.csv"
)

print("Dataset path:")
print(au_file_path)

Dataset path:
/content/drive/MyDrive/SE4050_DeepLearning/Lab_01/02_Datasets/Au_nanoparticle_dataset.csv


In [7]:
if os.path.exists(au_file_path):
    print("Au nanoparticle dataset found successfully.")
else:
    print("Dataset file was not found.")

Au nanoparticle dataset found successfully.


In [8]:
au_df = pd.read_csv(au_file_path)

print("Dataset loaded successfully.")
print("Dataset shape:", au_df.shape)

Dataset loaded successfully.
Dataset shape: (4000, 185)


In [9]:
au_df.head()

,ID,T,tau,time,N_total,N_bulk,N_surface,Volume,R_min,R_max,...,q6q6_S14,q6q6_S15,q6q6_S16,q6q6_S17,q6q6_S18,q6q6_S19,q6q6_S20,q6q6_S20+,Total_E,Formation_E
0,1,273,0.000025,1,1599,1014,585,2.710000e-26,15.5898,19.4516,...,0,0,0,0,0,0,0,0,-5908.6345,343.4555
1,2,273,0.000025,2,1642,1034,608,2.780000e-26,15.3996,24.1755,...,0,0,0,0,0,0,0,0,-6035.4034,384.8166
2,3,273,0.000025,3,4637,3365,1272,7.850000e-26,21.9199,31.0305,...,0,0,0,0,0,0,0,0,-17330.1660,800.5040
3,4,273,0.000025,4,7189,5292,1897,1.220000e-25,23.2542,37.8795,...,0,0,0,0,0,0,0,0,-26937.2010,1171.7890
4,5,273,0.000025,5,11004,8508,2496,1.860000e-25,27.1200,39.7614,...,0,0,0,0,0,0,0,0,-41427.9010,1597.7390


In [10]:
au_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Columns: 185 entries, ID to Formation_E
dtypes: float64(23), int64(162)
memory usage: 5.6 MB


In [11]:
print("Column names in the dataset:\n")

for number, column in enumerate(
    au_df.columns,
    start=1
):
    print(f"{number}. {repr(column)}")

Column names in the dataset:

1. 'ID'
2. 'T'
3. 'tau'
4. 'time'
5. 'N_total'
6. 'N_bulk'
7. 'N_surface'
8. 'Volume'
9. 'R_min'
10. 'R_max'
11. 'R_diff'
12. 'R_avg'
13. 'R_std'
14. 'R_skew'
15. 'R_kurt'
16. 'S_100'
17. 'S_111'
18. 'S_110'
19. 'S_311'
20. 'Curve_1-10'
21. 'Curve_11-20'
22. 'Curve_21-30'
23. 'Curve_31-40'
24. 'Curve_41-50'
25. 'Curve_51-60'
26. 'Curve_61-70'
27. 'Curve_71-80'
28. 'Curve_81-90'
29. 'Curve_91-100'
30. 'Curve_101-110'
31. 'Curve_111-120'
32. 'Curve_121-130'
33. 'Curve_131-140'
34. 'Curve_141-150'
35. 'Curve_151-160'
36. 'Curve_161-170'
37. 'Curve_171-180'
38. 'Avg_total'
39. 'Avg_bulk'
40. 'Avg_surf'
41. 'TCN_0'
42. 'TCN_1'
43. 'TCN_2'
44. 'TCN_3'
45. 'TCN_4'
46. 'TCN_5'
47. 'TCN_6'
48. 'TCN_7'
49. 'TCN_8'
50. 'TCN_9'
51. 'TCN_10'
52. 'TCN_11'
53. 'TCN_12'
54. 'TCN_13'
55. 'TCN_14'
56. 'TCN_15'
57. 'TCN_16'
58. 'TCN_17'
59. 'TCN_18'
60. 'TCN_19'
61. 'TCN_20'
62. 'BCN_0'
63. 'BCN_1'
64. 'BCN_2'
65. 'BCN_3'
66. 'BCN_4'
67. 'BCN_5'
68. 'BCN_6'
69. 'BCN_7'
70. '

In [12]:
au_df.columns = au_df.columns.str.strip()

print("Column names cleaned successfully.")

Column names cleaned successfully.


In [13]:
required_columns = [
    "N_total",
    "N_bulk",
    "N_surface",
    "R_avg"
]

missing_columns = [
    column
    for column in required_columns
    if column not in au_df.columns
]

if not missing_columns:
    print("All four required columns are available.")

    for column in required_columns:
        print("✓", column)
else:
    print("Missing columns:", missing_columns)

All four required columns are available.
✓ N_total
✓ N_bulk
✓ N_surface
✓ R_avg


## Task 2.2 — Select the Required Features

### Feature Selection

A new dataframe is created using only the four features required by the lab sheet:

- `N_total`
- `N_bulk`
- `N_surface`
- `R_avg`

The original dataframe remains unchanged.

In [14]:
new_df = au_df[
    [
        "N_total",
        "N_bulk",
        "N_surface",
        "R_avg"
    ]
].copy()

print("Filtered dataframe created successfully.")
print("Filtered dataframe shape:", new_df.shape)

Filtered dataframe created successfully.
Filtered dataframe shape: (4000, 4)


In [15]:
print("Filtered dataframe columns:")
print(new_df.columns.tolist())

Filtered dataframe columns:
['N_total', 'N_bulk', 'N_surface', 'R_avg']


In [16]:
print("Data types of selected features:\n")
print(new_df.dtypes)

Data types of selected features:

N_total        int64
N_bulk         int64
N_surface      int64
R_avg        float64
dtype: object


In [17]:
for column in required_columns:
    new_df[column] = pd.to_numeric(
        new_df[column],
        errors="coerce"
    )

print("Selected features converted to numeric format.")

Selected features converted to numeric format.


In [18]:
print("Missing values in selected features:\n")
print(new_df.isnull().sum())

Missing values in selected features:

N_total      0
N_bulk       0
N_surface    0
R_avg        0
dtype: int64


## Task 2.3 — Display the First 20 Samples

### First 20 Samples

The first 20 samples of the filtered dataframe are displayed below using the Pandas `head()` function.

In [19]:
first_20_samples = new_df.head(20)

print("First 20 samples of the selected features:")

display(first_20_samples)

First 20 samples of the selected features:


,N_total,N_bulk,N_surface,R_avg
0,1599,1014,585,17.3706
1,1642,1034,608,17.6061
2,4637,3365,1272,25.3692
3,7189,5292,1897,29.7011
4,11004,8508,2496,34.2831
5,13375,10768,2607,36.6334
6,13795,11155,2640,37.0108
7,13947,11304,2643,37.1672
8,14020,11357,2663,37.2103
9,14056,11389,2667,37.2467


In [20]:
first_20_display = (
    new_df
    .head(20)
    .copy()
)

first_20_display.index = range(
    1,
    len(first_20_display) + 1
)

first_20_display.index.name = "Sample Number"

display(first_20_display)

,N_total,N_bulk,N_surface,R_avg
Sample Number,,,,
1,1599,1014,585,17.3706
2,1642,1034,608,17.6061
3,4637,3365,1272,25.3692
4,7189,5292,1897,29.7011
5,11004,8508,2496,34.2831
6,13375,10768,2607,36.6334
7,13795,11155,2640,37.0108
8,13947,11304,2643,37.1672
9,14020,11357,2663,37.2103


In [21]:
print(
    "Number of displayed samples:",
    first_20_samples.shape[0]
)

print(
    "Number of displayed features:",
    first_20_samples.shape[1]
)

Number of displayed samples: 20
Number of displayed features: 4


In [22]:
print("Original dataframe shape:", au_df.shape)
print("Filtered dataframe shape:", new_df.shape)

print(
    "Original number of columns:",
    au_df.shape[1]
)

print(
    "Filtered number of columns:",
    new_df.shape[1]
)

Original dataframe shape: (4000, 185)
Filtered dataframe shape: (4000, 4)
Original number of columns: 185
Filtered number of columns: 4


### Interpretation

The Au nanoparticle dataset was loaded successfully using Pandas.

A new dataframe named `new_df` was created containing only the four required features: `N_total`, `N_bulk`, `N_surface`, and `R_avg`.

The filtered dataframe contains four columns, and its first 20 samples were displayed successfully. The original dataset remains unchanged.

In [23]:
# Task 2.1 to Task 2.3:
# Load the Au nanoparticle dataset,
# select the required features,
# and display the first 20 samples.

import os
import pandas as pd

dataset_folder = (
    "/content/drive/MyDrive/"
    "SE4050_DeepLearning/"
    "Lab_01/"
    "02_Datasets"
)

au_file_path = os.path.join(
    dataset_folder,
    "Au_nanoparticle_dataset.csv"
)

au_df = pd.read_csv(au_file_path)
au_df.columns = au_df.columns.str.strip()

selected_features = [
    "N_total",
    "N_bulk",
    "N_surface",
    "R_avg"
]

new_df = au_df[selected_features].copy()

for column in selected_features:
    new_df[column] = pd.to_numeric(
        new_df[column],
        errors="coerce"
    )

print("Original dataset shape:", au_df.shape)
print("Filtered dataframe shape:", new_df.shape)
print("\nFirst 20 samples:")

display(new_df.head(20))

Original dataset shape: (4000, 185)
Filtered dataframe shape: (4000, 4)

First 20 samples:


,N_total,N_bulk,N_surface,R_avg
0,1599,1014,585,17.3706
1,1642,1034,608,17.6061
2,4637,3365,1272,25.3692
3,7189,5292,1897,29.7011
4,11004,8508,2496,34.2831
5,13375,10768,2607,36.6334
6,13795,11155,2640,37.0108
7,13947,11304,2643,37.1672
8,14020,11357,2663,37.2103
9,14056,11389,2667,37.2467


In [24]:
assert new_df.shape[1] == 4, (
    "The filtered dataframe must contain exactly four columns."
)

assert list(new_df.columns) == required_columns, (
    "The selected columns are incorrect or in the wrong order."
)

assert len(new_df.head(20)) == 20, (
    "The output must contain exactly 20 samples."
)

print("All Phase 6 validation tests passed.")

All Phase 6 validation tests passed.


## Task 2.4 — Descriptive Statistical Analysis

## Task 2.5 — Histograms in a 1×4 Layout

## Task 2.6 — Seaborn Pairplot

## Task 2.7 — Customized Seaborn PairGrid